src/NHapi.Base/Model/AbstractGroup.cs

```diff
@@ -269,20 +256,37 @@ public virtual string AddNonstandardSegment(string name)
                 throw new HL7Exception("Need message version to add segment by name; message.getVersion() returns null");
             }
 
-            var c = myFactory.GetSegmentClass(name, version);
-            if (c == null)
+            var c = myFactory.GetSegmentClass(name, version) ?? typeof(GenericSegment);
+
+            var index = Names.Length;
+
+            TryToInstantiateStructure(c, name); // may throw exception
+
+            return Insert(c, false, true, false, index, name);
+        }
+
+        /// <inheritdoc />
+        public virtual string AddNonstandardSegment(string name, int index)
+        {
+            if (this is IMessage && index == 0)
             {
-                c = typeof(GenericSegment);
+                throw new HL7Exception($"Can not add nonstandard segment \"{name}\" to start of message.");

```

Comment:
can not -> cannot

also might look nice if you use single quotes in the message?
`throw new HL7Exception($"Cannot add nonstandard segment '{name}' to start of message.");`

Reply:
this was copied from hapi. 😂